Allow for the creation of unique subfolders in the `remote_workdir`

arosen93 opened the pull request:
Closes #56 and combines a separate bugfix I was going to introduce in #58.

```
### Added

- A new kwarg `create_unique_workdir` that will create unique subfolders of the type `<DISPATCH ID>/node_<NODE ID>` within `remote_workdir` if set to `True`


### Fixed

- Fixed a bug where `cleanup = False` would be ignored.
- Fixed a bug where if `cache_dir` was not present, Covalent would crash.
```

arosen93 commented:
@wjcunningham7, @santoshkumarradha, @cjao: I still have to thoroughly test this out, but how does this look for an implementation? Since we will want to mimic this approach in other executors, I want to make sure the name and overall approach is to your liking.

santoshkumarradha commented:
Wow this was super fast and perfect ! 🙏 

CC : @wjcunningham7

wjcunningham7 commented on covalent_slurm_plugin/slurm.py:
```diff
@@ -541,6 +558,9 @@ async def run(self, function: Callable, args: List, kwargs: Dict, task_metadata:
         node_id = task_metadata["node_id"]
         results_dir = task_metadata["results_dir"]
         task_results_dir = os.path.join(results_dir, dispatch_id)
+        self._current_remote_workdir = os.path.join(
```
I think this is really the only method that will end up using `current_remote_workdir` so you might want to make it a local variable here.

arosen93 replied:
It's used in a few places, so I've kept it under `self._`. Happy to adjust if you feel it necessary.

wjcunningham7 reviewed (changes requested):
Last comment is that the unique directory should be conveyed to Slurm using the `--chdir` option.

arosen93 commented on covalent_slurm_plugin/slurm.py:
```diff
@@ -121,6 +129,14 @@ def __init__(
 
         self.remote_workdir = remote_workdir or get_config("executors.slurm.remote_workdir")
 
+        if unique_workdir is None:
```
Keeping this one as-is for the same reason as the cleanup bug. If set to False, the config value is always used.

arosen93 commented on covalent_slurm_plugin/slurm.py:
```diff
@@ -163,7 +184,10 @@ def __init__(
             self.poll_freq = 60
 
-        self.cleanup = cleanup or get_config("executors.slurm.cleanup")
+        if cleanup is None:
```
Discussed over Slack. If cleanup is False, the config value is used.

arosen93 commented on covalent_slurm_plugin/slurm.py:
```diff
@@ -549,11 +569,11 @@ async def run(self, function: Callable, args: List, kwargs: Dict, task_metadata:
         if "output" not in self.options:
             self.options["output"] = os.path.join(
-                self.remote_workdir, f"stdout-{dispatch_id}-{node_id}.log"
+                self._current_remote_workdir, f"stdout-{dispatch_id}-{node_id}.log"
```
Discussed over Slack. We will keep this one as-is. And I think that's a good decision because a lot of code that I use actually relies on the stdout and stderr being in the current working directory.

arosen93 commented:
This is good to go from my side but a few minor comments from my side:

- My personal opinion is that we should remove the `source $HOME/.bashrc` line. If the user needs to source their .bashrc, they should include that as a prerun_command. We don't want to force the sourcing of the .bashrc (for instance, on NERSC, there are several .bashrc files and the usual $HOME/.bashrc isn't always the desired one). I've removed the line altogether, but if you disagree, let me know.
- If the `cache_dir` doesn't exist, Covalent crashes. I've fixed that now.
